htlcswitch/switch_test.go

```diff
@@ -3318,3 +3320,370 @@ func TestSwitchHoldForward(t *testing.T) {
 	assertNumCircuits(t, s, 0, 0)
 }
+
+// TestSwitchDustForwarding tests that the switch properly fails HTLC's which
+// have incoming or outgoing links that breach their dust thresholds.
+func TestSwitchDustForwarding(t *testing.T) {
+	t.Parallel()
+
+	// We'll create a three-hop network:
+	// - Alice has a dust limit of 200sats with Bob
+	// - Bob has a dust limit of 800sats with Alice
+	// - Bob has a dust limit of 200sats with Carol
+	// - Carol has a dust limit of 800sats with Bob
+	channels, cleanUp, _, err := createClusterChannels(
+		btcutil.SatoshiPerBitcoin, btcutil.SatoshiPerBitcoin,
+	)
+	require.NoError(t, err)
+	defer cleanUp()
+
+	n := newThreeHopNetwork(
+		t, channels.aliceToBob, channels.bobToAlice,
+		channels.bobToCarol, channels.carolToBob, testStartingHeight,
+	)
+	err = n.start()
+	require.NoError(t, err)
+
+	// We'll also put Alice and Bob into hodl.ExitSettle mode, such that
+	// they won't settle incoming exit-hop HTLC's automatically.
+	n.aliceChannelLink.cfg.HodlMask = hodl.ExitSettle.Mask()
+	n.firstBobChannelLink.cfg.HodlMask = hodl.ExitSettle.Mask()
+
+	// We'll test that once the default threshold is exceeded on the
+	// Alice -> Bob channel, either side's calls to SendHTLC will fail.
+	// This does not rely on the mailbox sum since there's no intermediate
+	// hop.
+	//
+	// Alice will send 357 HTLC's of 700sats. Bob will also send 357 HTLC's
+	// of 700sats. If either side attempts to send a dust HTLC, it will
+	// fail so amounts below 800sats will breach the dust threshold.
+	aliceCount := 357
+	bobCount := 357
+	amt := lnwire.NewMSatFromSatoshis(700)
+
+	// Create the hops that Alice will use to send payments to Bob.
+	htlcAmt, totalTimelock, aliceHops := generateHops(
+		amt, testStartingHeight, n.firstBobChannelLink,
+	)
+	firstHop := n.aliceChannelLink.ShortChanID()
+
+	// Convert the hops to a blob that will be put in the Add message.
+	aliceBlob, err := generateRoute(aliceHops...)
+	require.NoError(t, err)
+
+	// Create a slice to store the preimages used in making Alice's dust
+	// payments to Bob.
+	alicePreimages := make([]lntypes.Preimage, aliceCount)
+
+	// Initialize alice's attempt ID.
+	aliceAttemptID := 0
+
+	// Generate Alice's preimages that will be used to send hold payments.
+	for i := 0; i < aliceCount; i++ {
+		// Deterministically generate preimages. Avoid the all-zeroes
+		// preimage because that will be rejected by the database.
+		alicePreimages[i] = lntypes.Preimage{byte(i >> 8), byte(i), 1}
+	}
+
+	// Alice calls SendHTLC in a loop for aliceCount HTLC's.
+	for i := 0; i < aliceCount; i++ {
+		// Construct the htlc packet.
+		hash := alicePreimages[i].Hash()
+
+		htlc := &lnwire.UpdateAddHTLC{
+			PaymentHash: hash,
+			Amount:      htlcAmt,
+			Expiry:      totalTimelock,
+			OnionBlob:   aliceBlob,
+		}
+
+		err = n.aliceServer.htlcSwitch.SendHTLC(
+			firstHop, uint64(aliceAttemptID), htlc,
+		)
+		require.NoError(t, err)
+		aliceAttemptID++
+	}
```

Comment:
Could pull all of this out into a helper since we duplicate it for bob.

Reply:
good point

Reply:
done, PTAL